doc/build-file-of-options.1
.TH BUILD\-FILE\-OF\-OPTIONS 1 "August 2023" "build-file-of-options 1.x" "User Commands"
.SH NAME
build\-file\-of\-options \- Build a list of command line options from a .conf file
.SH SYNOPSIS
.B build\-file\-of\-options
[\fIOPTION\fR]... \-\-output \fI<filename>.cpp\fR \fI<filename>.conf\fR

.SH DESCRIPTION
This command reads a Unix configuration file and parses the comments for
command line option definitions.
.PP
The first comment at the top of the file, up to the first empty line, is
always discarded.
.PP
The following commands must start with a `<name>=<args>' parameter definition
to be detected as a command line option. The first sentence found after the
command line parameter (i.e. up to the first period: `.') is taken as the
help string. The comment can end with a `Default: <value>'.

.SH "NAME"
A command name cannot include spaces, colons, or equal signs.
.PP
If the name includes a `|' character, then the name in split in two. The
left part is the official command line option. The second part is an alias.
For example, you could define `licence|license' to support the English and
French spellings of the word licence.

.SH "SHORT NAME"
If the option parameter ends with a letter between parenthesis, then it is
taken as the short name. For example, `--input=filename (i)' defines an
option named "input" with the short name set to "i".

.B IMPORTANT:
The short name must be exactly one letter. It can be nearly any Unicode
character (diacriticals, such as character code 0x0300, are not valid).

.SH "ARGUMENTS"
The option name must be followed by an equal sign. You can then specify a
list of possible arguments. It may be as simple as `--output=<filename>'.
.PP
The arguments can be made optional by enclosing them between square brackets.
For example, the `color=[red|green|blue]' means that the `\-\-color' parameter
can be used by itself and a default color will be used.
.PP
Similarly, the arguments can be marked as supporting multiple entries instead
of just one. This is doable by either putting the parameter between curvly
braces or by ending the list of arguments with an elipsis: `...'. For example,
a list of network ports could be defined as `ports={<number from 1 to 65535>}'.
.PP
Arguments that are optional and can be repeated must use the elipsis. For
example, the `height=[<number> ...]' parameter can be specified by itself
(`\-\-height'), with one number (`\-\-height 123'), or multiple numbers
(`\-\-height 123 456 789').

.SH "DEFAULT VALUE"
If your parameter supports a default value, you can specific that value
using a line after the help that starts with the word `Default:'. Whatever
appears after that (trimmed), is saved as the default.

.B Note:
An empty string is not currently supported as a default value.

.SH "COMMAND LINE OPTIONS"
.TP
\fB\-\-build\-date\fR
Print the date when this tool was compiled.

.TP
\fB\-\-compiler\-version\fR
Print the version of the C++ compiler used to compile this tool.

.TP
\fB\-\-configuration\-filenames\fR
Print the list of files searched to configure this tool.

.TP
\fB\-C\fR, \fB\-\-copyright\fR
Print the copyright notice of the `build\-file\-of\-options' tool.

.TP
\fB\-\-environment\-variable\-name\fR
Print the name of the environment variable that can be used to pass arguments
to this tool.

.TP
\fB\-\-has\-sanitizer\fR
Print whether the tool was compiled with the sanitizer options.

.TP
\fB\-h\fR, \fB\-\-help\fR
Print a brief document about the tool usage, then exit.

.TP
\fB\-L\fR, \fB\-\-license\fR
Print out the license of `build\-file\-of\-options' and exit.

.TP
\fB\-?\fR, \fB\-\-long\-help\fR
Print out all the available command line options.

.TP
\fB\-o\fR, \fB\-\-output\fR \fIFILENAME\fR
Define the name of the file where the results are saved. \fIFILENAME\fR
may be set to `-' to output the data to stdout (default if not specified).

.TP
\fB\-\-path\-to\-option\-definitions\fR
Option definitions can be defined in a .ini file. If it exists, this is the
path where it can be found.

.TP
\fB\-\-print\-option\fR \fINAME\fR
By default, the command keeps the comments found in the original file.
This can cause some issues as the output is sorted by known fields.
If that's the case with the file you are dealing with, use this
option to remove all the comments and only keep field names and
value pairs.

.TP
\fB\-\-show\-option\-sources\fR
The `advgetopt' library has the ability to trace where each value is
read from: which file, the command line, the environment variable.
This option can be used to determine where a value is defined, which
once in a while is particularly useful.

.TP
\fB\-S\fR, \fB\-\-show\-system\-help\fR
Print the list of system command line options.

.TP
\fB\-V\fR, \fB\-\-version\fR
Print version number, then exit

.SH AUTHOR
Written by Alexis Wilke <alexis@m2osw.com>.
.SH "REPORTING BUGS"
Report bugs to <https://github.com/m2osw/advgetopt/issues>.
.br
advgetopt home page: <https://snapwebsites.org/project/advgetopt>.
.SH COPYRIGHT
Copyright \(co 2022-2025  Made to Order Software Corp.  All Rights Reserved
.br
License: GPLv3
.br
This is free software: you are free to change and redistribute it.
.br
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR atomic-names(1)


debian/changelog
advgetopt (2.0.47.0~jammy) jammy; urgency=high

  * Added a function to reset the configuration file cache.

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 10 Dec 2023 15:07:57 -0800

advgetopt (2.0.46.1~bionic) bionic; urgency=high

  * Bumped build version to rebuild on Launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Fri, 10 Nov 2023 19:21:40 -0800

advgetopt (2.0.46.0~jammy) jammy; urgency=high

  * Added missing #include <cstdint>

 -- Alexis Wilke <alexis@m2osw.com>  Tue, 07 Nov 2023 05:15:44 +0000

advgetopt (2.0.45.0~jammy) jammy; urgency=high

  * Added two more sizes (ronna & quetta).
  * Automatically refuse to start if code 9 is returned.
  * Allow '\0' to be appended to a string.
  * Updated debian/compat to 15.
  * Implemented a first version and install the option generator.
  * Replaced random_shuffle() with just shuffle().

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 05 Nov 2023 21:24:46 -0800

advgetopt (2.0.44.0~bionic) bionic; urgency=high

  * Added the missing option_keys param documentation.
  * Added option to support any configuration setup for standard config files.
  * Implemented operator support in advgetopt_config. This requires the setup
    with the EXTENDED variable support (hence the additional of config_setup).

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 12 Nov 2022 15:56:47 -0800

advgetopt (2.0.43.0~bionic) bionic; urgency=high

  * Added a "show system" flag so we can hide system options by default.
  * Enhanced Help() from --print-option.
  * Changed when the parse_options_from_group_names() gets called.
  * Avoid the double error in atomic-names tool on command line errors.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 12 Nov 2022 11:00:03 -0800

advgetopt (2.0.42.0~bionic) bionic; urgency=high

  * Allow for empty values in the ranges representing min/max.
  * Use snapdev::trim_string() instead of the boost version.

 -- Alexis Wilke <alexis@m2osw.com>  Fri, 11 Nov 2022 21:06:04 -0800

advgetopt (2.0.41.0~bionic) bionic; urgency=high

  * Added the validator_length validator.
  * Added support for option keys (ARRAY flag).
  * Added support for the +=, ?=, and := assignment operators.
  * Added a test to verify the basics with the assignment operators.
  * Removed the need for boost library from one more file.
  * Removed the operator+ of char32_t/string (see libutf8).
  * Show aliases in the usage output (i.e. --help screen).
  * Made the function converting from to '-'/'_' available as an util.
  * Fixed a crash on a nullptr variable pointer.
  * Fixed the generation of filenames and added a clear priority scheme.
  * Made add_option_from_string() public (for fluid-settings).
  * Made get_options_filename() public.
  * Support the "dynamic-configuration" flag in "allow=...".
  * Added a "...::group" parameter to group options defined in a file.
  * Added the "length" validator.

 -- Alexis Wilke <alexis@m2osw.com>  Fri, 15 Jul 2022 09:02:06 -0700

advgetopt (2.0.40.0~bionic) bionic; urgency=high

  * Removed the debugs. The g_validator_factories was the culprit.

 -- Alexis Wilke <alexis@m2osw.com>  Thu, 14 Jul 2022 21:17:18 -0700

advgetopt (2.0.39.3~bionic) bionic; urgency=high

  * Changed the g_validator_factories in a pointer so we allocate before using.

 -- Alexis Wilke <alexis@m2osw.com>  Thu, 14 Jul 2022 20:20:45 -0700

advgetopt (2.0.39.2~bionic) bionic; urgency=high

  * Try running the software with gdb to see if we can get some info about
    the crash. If it happens before main()...

 -- Alexis Wilke <alexis@m2osw.com>  Thu, 14 Jul 2022 15:06:44 -0700

advgetopt (2.0.39.1~bionic) bionic; urgency=high

  * Added debug code inline since I don't see the signal handler message.

 -- Alexis Wilke <alexis@m2osw.com>  Thu, 14 Jul 2022 11:52:04 -0700

advgetopt (2.0.39.0~bionic) bionic; urgency=high

  * Use a signal handler to capture crashes and print stack in atomic-names.

 -- Alexis Wilke <alexis@m2osw.com>  Wed, 13 Jul 2022 21:57:28 -0700

advgetopt (2.0.38.1~bionic) bionic; urgency=high

  * Applied fix to a static_cast so it compiles in release.

 -- Alexis Wilke <alexis@m2osw.com>  Mon, 11 Jul 2022 10:02:59 -0700

advgetopt (2.0.38.0~bionic) bionic; urgency=high

  * Fixed two bugs in the environment variable check function.
  * Properly define f_variables in all cases.
  * When parsing environment variable, distinguish between empty and undefined.
  * Allow for flags (i.e. "--verbose") to accept "true" and "false".
  * Added '=', '.', and '/' to the list of characters not requiring escaping.
  * Added coverage tests for new code (validators, variables, etc.)
  * Applied some fixes to the variables implementation so errors work.
  * Fixed the search of the atomic-names tool when building coverage.
  * Also generate a fatal error if we can't find the tool at all.
  * Added the keywords & list validators to make it possible to verify things
    such as "off,min,max" or an integer.
  * If the output of --help is longer than one screen, use `less`.

 -- Alexis Wilke <alexis@m2osw.com>  Thu, 26 May 2022 21:42:41 -0700

advgetopt (2.0.37.1~bionic) bionic; urgency=high

  * Added two echos around the atomic program to test a crash on launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 21 May 2022 21:46:36 -0700

advgetopt (2.0.37.0~bionic) bionic; urgency=high

  * The first find_program() of atomic-names must not be REQUIRED.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 21 May 2022 11:05:50 -0700

advgetopt (2.0.36.0~bionic) bionic; urgency=high

  * Added atomic-names tool to -dev package.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 21 May 2022 08:16:46 -0700

advgetopt (2.0.35.1~bionic) bionic; urgency=high

  * Bumped build version to rebuild on Launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Fri, 20 May 2022 12:57:33 -0700

advgetopt (2.0.35.0~bionic) bionic; urgency=high

  * Added tool/cmake rule to automatically generate Atomic Names.
  * Added a exists() function for configuration files.
  * Fixes to the CMakeLists.txt files.

tools/build-file-of-options.cpp
// Copyright (c) 2020-2022  Made to Order Software Corp.  All Rights Reserved
//
// https://snapwebsites.org/project/advgetopt
// contact@m2osw.com
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


/** \file
 * \brief build-file-of-options tool.
 *
 * We use this tool to convert the comments found in our configuration files
 * in a list of options that can be parsed by the advgetopt objects.
 *
 * Specifically, it understands the option name, default value, and when
 * available some other features such as ranges and types.
 */


// advgetopt lib
//
#include    <advgetopt/advgetopt.h>
#include    <advgetopt/exception.h>
#include    <advgetopt/version.h>


// snapdev lib
//
#include    <snapdev/not_reached.h>


// boost
//
#include    <boost/preprocessor/stringize.hpp>


// C++ lib
//
#include    <iostream>
#include    <fstream>
#include    <sstream>


// last include
//
#include    <snapdev/poison.h>




namespace
{

/** \brief Command line options.
 *
 * This table includes all the command line options supported by the
 * `build-file-of-options` tool.
 */
advgetopt::option const g_options[] =
{
    advgetopt::define_option(
          advgetopt::Name("output")
        , advgetopt::ShortName('o')
        , advgetopt::Flags(advgetopt::command_flags<advgetopt::GETOPT_FLAG_REQUIRED>())
        , advgetopt::Help("Specify the path and filename of the output file.")
    ),
    advgetopt::define_option(
          advgetopt::Name("verbose")
        , advgetopt::ShortName('v')
        , advgetopt::Flags(advgetopt::option_flags<advgetopt::GETOPT_FLAG_COMMAND_LINE
                                                 , advgetopt::GETOPT_FLAG_ENVIRONMENT_VARIABLE
                                                 , advgetopt::GETOPT_FLAG_CONFIGURATION_FILE>())
        , advgetopt::Help("Show commands being executed.")
    ),
    advgetopt::define_option(
          advgetopt::Name("--")
        , advgetopt::Flags(advgetopt::command_flags<advgetopt::GETOPT_FLAG_DEFAULT_OPTION>())
        , advgetopt::Help("<configuration filename>")
    ),
    advgetopt::end_options()
};


/** \brief The tool looks for this configuration file.
 *
 * The build-file-of-options allow you to have a configuration file
 * with various options in it.
 */
constexpr char const * const g_configuration_files[]
{
    "/etc/advgetopt/build-file-of-options.conf",
    nullptr
};




// TODO: once we have stdc++20, remove all defaults
#pragma GCC diagnostic ignored "-Wpedantic"
advgetopt::options_environment const g_options_environment =
{
    .f_project_name = "build-file-of-options",
    .f_group_name = nullptr,
    .f_options = g_options,
    .f_options_files_directory = nullptr,
    .f_environment_variable_name = "BUILD_FILE_OF_OPTIONS_OPTIONS",
    .f_section_variables_name = nullptr,
    .f_configuration_files = g_configuration_files,
    .f_configuration_filename = nullptr,
    .f_configuration_directories = nullptr,
    .f_environment_flags = advgetopt::GETOPT_ENVIRONMENT_FLAG_PROCESS_SYSTEM_PARAMETERS,
    .f_help_header = "Usage: %p [-<opt>] <configuration file>\n"
                     "where -<opt> is one or more of:",
    .f_help_footer = "%c",
    .f_version = LIBADVGETOPT_VERSION_STRING,
    .f_license = nullptr,
    .f_copyright = "Copyright (c) 2019-" BOOST_PP_STRINGIZE(UTC_BUILD_YEAR) "  Made to Order Software Corporation",
    .f_build_date = UTC_BUILD_DATE,
    .f_build_time = UTC_BUILD_TIME
};





class build_file
{
public:
                                    build_file(int argc, char * argv[]);

    int                             run();

private:
    advgetopt::getopt               f_opt;
};



build_file::build_file(int argc, char * argv[])
    : f_opt(g_options_environment, argc, argv)
{
    if(f_opt.is_defined("help"))
    {
        std::cout << f_opt.usage();
        exit(1);
    }

    if(f_opt.is_defined("version"))
    {
        std::cout << LIBADVGETOPT_VERSION_STRING << std::endl;
        exit(1);
    }
}


int build_file::run()
{

// TODO: actually implement the tool!

    return 0;
}




} // no name namespace




int main(int argc, char * argv[])
{
    try
    {
        build_file session(argc, argv);
        return session.run();
    }
    catch(advgetopt::getopt_exit const & e)
    {
        exit(e.code());
    }
    catch(std::exception const & e)
    {
        std::cerr << "error: exception caught: " << e.what() << std::endl;
        return 1;
    }
    snap::NOT_REACHED();
}


// vim: ts=4 sw=4 et

